truffle/com.oracle.truffle.api.object.dsl/src/com/oracle/truffle/api/object/dsl/Layout.java

```diff
@@ -0,0 +1,412 @@
+ * when creating new instances. As shape properties are set and changed, multiple factories will be
+ * created and it is up to the user to store and supply these as needed.
+ *
+ * Consider the example of a Java-style object, with a class and a hash code. The class would be a
+ * shape property, as many objects will share the same class, and the hash code will be a normal
+ * property.
+ *
+ * Shape properties are created by parameters in the method that creates the shape. The factory
+ * method then accepts an instance of a factory when creating the object, which is how the instance
+ * knows the value of the class property to use. A getter for a shape property can be defined as
+ * normal.
+ *
+ * {@codesnippet javaobject}
+ *
+ * When we load our Java interpreter we need to set the class property of the {@code Class} object
+ * to be itself. This means in this one isolated, slow-path, case we need to change a shape property
+ * for an object that is already allocated. Getters for shape properties can be defined for the
+ * {@link DynamicObjectFactory}, and for the {@link ObjectType}.
+ *
+ * Setters for shape properties are more complex, and they are not intended to be used in the fast
+ * path. Setters can be defined on a {@link DynamicObjectFactory}, in which case they return a new
+ * factory, or on a {@link DynamicObject}, in which they they change the shape of the object. This
+ * is a slow-path operation and is likely to invalidate caches in your interpreter.
+ *
+ * {@codesnippet shapesetters}
+ *
+ * Apply this to our example with Java classes:
+ *
+ * {@codesnippet closecycle}
+ *
+ * <h2>Layout Inheritance</h2>
+ *
+ * Inheritance of layout interfaces allows you to model classical class inheritance, such as in a
+ * language like Java. Use normal interface inheritance to make one layout inherit from another. You
+ * then need to add the parameters for super-layouts at the beginning of sub-layout constructor
+ * methods.
+ *
+ * Inherited shape properties work in a similar way.
+ *
+ * {@codesnippet inheritanceinterfaces}
+ *
+ * {@codesnippet inheritanceuse}
+ *
+ * <h2>Custom Object-Type Superclass</h2>
+ *
+ * Generated layouts use custom {@link ObjectType} subclasses internally. The default base class
+ * that is inherited from is simply {@link ObjectType}. You can change this with the
+ * {@link #objectTypeSuperclass} property on the {@link Layout} annotation.
+ *
+ * <h2>Custom Identifiers</h2>
+ *
+ * By default, internal {@link HiddenKey} identifiers with descriptive names will be created for
+ * your properties automatically. You can also specify the identifiers to use by defining them as
+ * public static final fields in the interface.
+ *
+ * {@codesnippet customid}
+ */
+@Target(ElementType.TYPE)
+@Retention(RetentionPolicy.SOURCE)
+public @interface Layout {
+
+    Class<? extends ObjectType> objectTypeSuperclass() default ObjectType.class;
+
+    class Snippets {
```

Comment:
Every member of an interface is public, so you can't hide Snippets at this place. But you could move the `class Snippets` definition to the end of this file. A .java file can contain many subsequent class definitions. At most one of them can be public. So if you add `class LayoutSnippets` to the end (or even beginning) of this file it should be gone from the Javadoc.

If we agree on same naming scheme, we can even modify mx to not include **/*Snippet*.class files in the final JAR. With that in place most of the issues raised in #19 might be solved without negative impact on the Javadoc and final JAR bits.